lib/min_heap.rb

```diff
@@ -44,28 +53,55 @@ def to_s
   end
 
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(n)
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    return @store.length == 0 
   end
 
   private
 
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(log n)
+  # Space complexity: O(1)
   def heap_up(index)
-    
+    # return if there's only one node 
+    return if index == 0 
+    # root node is...
+    parent = (index-1)/2
+    # if node is greater than root, it's in order, do nothing 
+    return if @store[index].key > @store[parent].key
+    # otherwise, swap the two
+    swap(index, parent)
+    # check again 
+    heap_up(parent)
   end
 
   # This helper method takes an index and 
   #  moves it up the heap if it's smaller
   #  than it's parent node.
   def heap_down(index)
-    raise NotImplementedError, "Method not implemented yet..."
+    return if @store.empty?
+
+    left_child = 2*index+1
+    right_child = 2*index+2
+
+    # if either is nil, can use, so return
+    return if @store[left_child].nil? || @store[right_child].nil?
```

Comment:
What if the left child is not `nil`, but it **IS** smaller than the parent?  Small edge-case the test missed here.